File: DAA_LAB/Experiment_6_DisjointSetUnion/DisjointSetUnion.py
import random
import time

parent = []
rank = []

def make_set(n):

    global parent, rank

    parent = [i for i in range(n)]
    rank = [0] * n

def find(x):

    if parent[x] != x:
        parent[x] = find(parent[x])

    return parent[x]

def union(x, y):

    root_x = find(x)
    root_y = find(y)

    if root_x == root_y:
        return

    if rank[root_x] < rank[root_y]:

        parent[root_x] = root_y

    elif rank[root_x] > rank[root_y]:

        parent[root_y] = root_x

    else:

        parent[root_y] = root_x
        rank[root_x] += 1


n = 1000

make_set(n)

start = time.time()

for _ in range(500):

    a = random.randint(0, n - 1)
    b = random.randint(0, n - 1)

    union(a, b)

end = time.time()

print("Friend Groups:")

for i in range(10):

    print(
        f"User {i} -> Group {find(i)}"
    )

print()

print("Execution Time =",
      end - start,
      "seconds")

File: DAA_LAB/Experiment_6_DisjointSetUnion/DisjointSetUnion.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define MAX 10000

int parent[MAX];
int rankArr[MAX];

void makeSet(int n)
{
    for(int i = 0; i < n; i++)
    {
        parent[i] = i;
        rankArr[i] = 0;
    }
}

int find(int x)
{
    if(parent[x] != x)
        parent[x] = find(parent[x]);

    return parent[x];
}

void unionSets(int x, int y)
{
    int rootX = find(x);
    int rootY = find(y);

    if(rootX == rootY)
        return;

    if(rankArr[rootX] < rankArr[rootY])
        parent[rootX] = rootY;

    else if(rankArr[rootX] > rankArr[rootY])
        parent[rootY] = rootX;

    else
    {
        parent[rootY] = rootX;
        rankArr[rootX]++;
    }
}

int main()
{
    int n = 1000;

    srand(time(NULL));

    makeSet(n);

    clock_t start = clock();

    for(int i = 0; i < 500; i++)
    {
        int a = rand() % n;
        int b = rand() % n;

        unionSets(a, b);
    }

    clock_t end = clock();

    double executionTime =
        (double)(end - start) /
        CLOCKS_PER_SEC;

    printf("Friend Groups:\n");

    for(int i = 0; i < 10; i++)
        printf("User %d -> Group %d\n",
               i,
               find(i));

    printf("\nExecution Time = %lf seconds\n",
           executionTime);

    return 0;
}
